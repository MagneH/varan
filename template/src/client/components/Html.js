import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import serialize from 'serialize-javascript';

class Html extends PureComponent {
  static propTypes = {
    title: PropTypes.node,
    meta: PropTypes.node,
    link: PropTypes.node,
    style: PropTypes.node,
    script: PropTypes.node,
    noscript: PropTypes.node,
    base: PropTypes.node,
    htmlAttributes: PropTypes.object,
    bodyAttributes: PropTypes.object,
    bundleJs: PropTypes.arrayOf(PropTypes.string.isRequired),
    bundleCss: PropTypes.arrayOf(PropTypes.string.isRequired),
    manifest: PropTypes.string,
    body: PropTypes.string,
    initialState: PropTypes.object,
    preload: PropTypes.arrayOf(PropTypes.string.isRequired),
  };
  static defaultProps = {
    htmlAttributes: {},
    bodyAttributes: {},
    body: '',
    preload: [],
  };
  render() {
    const {
      htmlAttributes,
      bodyAttributes,
      title,
      meta,
      link,
      style,
      script,
      noscript,
      base,
      body,
      bundleJs,
      bundleCss,
      initialState,
      manifest,
      preload,
    } = this.props;
    return (
      <html {...htmlAttributes}>
        <head>
          {title}
          {meta}
          {noscript}
          {base}
          {manifest && <link rel="manifest" href={manifest} />}
          {link}
          {preload.map((file, i) => {
            if (/\.js$/.test(file)) return <link key={i} href={file} rel="preload" as="script" />;
            if (/\.css$/.test(file)) return <link key={i} href={file} rel="preload" as="style" />;
            if (/(\.woff|\.woff2|\.eot|\.ttf)$/.test(file))
              return <link key={i} href={file} rel="preload" as="font" crossOrigin="anonymous" />;
            if (/(\.png|\.jpe?g|\.gif)$/.test(file))
              return <link key={i} href={file} rel="preload" as="image" crossOrigin="anonymous" />;
            return null;
          })}
          {bundleCss.map((css, i) => (
            <link key={i} href={css} rel="stylesheet" />
          ))}
          {style}
          {script}
        </head>
        <body {...bodyAttributes}>
          <div id="root" dangerouslySetInnerHTML={{ __html: body }} />
          {initialState && (
            <script
              id="initial-state"
              type="text/javascript"
              dangerouslySetInnerHTML={{
                __html: `window.__INITIAL_REDUX_STATE__ = ${serialize(initialState, { isJSON: true })}`,
              }}
            />
          )}
          {bundleJs.map((js, i) => (
            <script key={i} type="text/javascript" src={js} defer />
          ))}
        </body>
      </html>
    );
  }
}

export default Html;
