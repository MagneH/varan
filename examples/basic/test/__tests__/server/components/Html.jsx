import React from 'react';
import renderer from 'react-test-renderer';
import Html from '../../../../src/server/components/Html';

// Tests
test('renders correctly', () => {
  const component = renderer.create(
    <Html
      htmlAttributes={{ id: 'html-id' }}
      bodyAttributes={{ id: 'body-id' }}
      title={<title>title</title>}
      meta={[<meta charSet="utf-8" />, <meta name="description" content="A varan react starter app" />]}
      link={[
        <link href="style.css" rel="stylesheet" />,
        <link rel="icon" href="favicon.ico" data-react-helmet="true" />,
      ]}
      style={<style>{'html { background-color: red; }'}</style>}
      script={[
        <script id="somescript" type="text/javascript">
          {'window.asd = true;'}
        </script>,
      ]}
      noscript={<noscript>javascript is required</noscript>}
      base={<base target="_blank" href="http://example.com/" />}
      body={<p>the html body</p>}
      bundleJs={['main.js', 'extra.js']}
      bundleCss={['main.css', 'extra.css']}
      manifest="manifest.json"
      preload={['jsfile.js', 'cssfile.css', 'fontfile.wolff2', 'imgfile.png']}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
