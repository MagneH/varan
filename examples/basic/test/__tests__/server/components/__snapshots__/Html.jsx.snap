// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<html
  id="html-id"
  lang="en"
>
  <head>
    <title>
      title
    </title>
    <meta
      charSet="utf-8"
    />
    <meta
      content="A varan react starter app"
      name="description"
    />
    <noscript>
      javascript is required
    </noscript>
    <base
      href="http://example.com/"
      target="_blank"
    />
    <link
      href="manifest.json"
      rel="manifest"
    />
    <link
      href="style.css"
      rel="stylesheet"
    />
    <link
      data-react-helmet="true"
      href="favicon.ico"
      rel="icon"
    />
    <link
      as="script"
      href="jsfile.js"
      rel="preload"
    />
    <link
      as="style"
      href="cssfile.css"
      rel="preload"
    />
    <link
      as="image"
      crossOrigin="anonymous"
      href="imgfile.png"
      rel="preload"
    />
    <link
      href="main.css"
      rel="stylesheet"
    />
    <link
      href="extra.css"
      rel="stylesheet"
    />
    <style>
      html { background-color: red; }
    </style>
    <script
      id="somescript"
      type="text/javascript"
    >
      window.asd = true;
    </script>
  </head>
  <body
    id="body-id"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": <p>
            the html body
          </p>,
        }
      }
      id="root"
    />
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "window.__INITIAL_REDUX_STATE__ = {}",
        }
      }
      id="initial-state"
      type="text/javascript"
    />
    <script
      defer={true}
      src="main.js"
      type="text/javascript"
    />
    <script
      defer={true}
      src="extra.js"
      type="text/javascript"
    />
  </body>
</html>
`;
